### PR TITLE
Implementation of Chef External Secrets Provider

### PR DESCRIPTION
**Pull request by sourav977**
## Problem Statement

Introducing Chef External Secrets Provider. 
This will enable users to seamlessly integrate Chef-based secret management with Kubernetes through the existing External Secrets framework.

## Related Issue

issue: https://github.com/external-secrets/external-secrets/issues/2905

## Related work
- documentation: docs/provider/chef.md
- api: apis/externalsecrets/v1beta1/secretstore_chef_types.go
- implementation: pkg/provider/chef

## Checklist

- [x] I have read the [contribution guidelines](https://external-secrets.io/latest/contributing/process/#submitting-a-pull-request)
- [x] All commits are signed with `git commit --signoff`
- [x] My changes have reasonable test coverage
- [x] All tests pass with `make test`
- [x] I ensured my PR is ready for review with `make reviewable`

### CODE REVIEW

File: apis/externalsecrets/v1beta1/secretstore_chef_types.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+// ChefProvider configures a store to sync secrets using basic chef server connection credentials.
+type ChefProvider struct {
+	// Auth defines the information necessary to authenticate against chef Server
+	Auth *ChefAuth `json:"auth"`
+	// UserName should be the user ID on the chef server
+	UserName string `json:"username"`
+	// ServerURL is the chef server URL used to connect to. If using orgs you should include your org in the url and terminate the url with a "/"
+	ServerURL string `json:"serverurl"`
```
Comment (Skarlso) [inline apis/externalsecrets/v1beta1/secretstore_chef_types.go:37]:
```suggestion
	ServerURL string `json:"serverUrl"`
```

or

```suggestion
	ServerURL string `json:"serverURL"`
```

File: pkg/provider/chef/chef.go [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+	"github.com/external-secrets/external-secrets/apis/externalsecrets/v1beta1"
+	"github.com/external-secrets/external-secrets/pkg/metrics"
+	"github.com/external-secrets/external-secrets/pkg/utils"
+)
+
+const (
+	errChefStore                             = "received invalid Chef SecretStore resource: %w"
+	errMissingStore                          = "missing store"
+	errMisingStoreSpec                       = "missing store spec"
```
Comment (Skarlso) [inline pkg/provider/chef/chef.go:39]:
```suggestion
	errMissingStoreSpec                       = "missing store spec"
```

File: pkg/provider/chef/chef.go [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+		if chefProvider.Auth.SecretRef.SecretKey.Namespace == nil {
+			return nil, fmt.Errorf(errInvalidClusterStoreMissingPKNamespace)
+		}
+		objectKey.Namespace = *chefProvider.Auth.SecretRef.SecretKey.Namespace
+	}
+
+	err = kube.Get(ctx, objectKey, credentialsSecret)
+	if err != nil {
+		return nil, fmt.Errorf(errFetchK8sSecret, err)
+	}
```
Comment (Skarlso) [inline pkg/provider/chef/chef.go:115]:
```suggestion
	if err := kube.Get(ctx, objectKey, credentialsSecret); err != nil {
		return nil, fmt.Errorf(errFetchK8sSecret, err)
	}
```

File: pkg/provider/chef/chef.go [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+
+	secretKey := credentialsSecret.Data[chefProvider.Auth.SecretRef.SecretKey.Key]
+	if (secretKey == nil) || (len(secretKey) == 0) {
```
Comment (Skarlso) [inline pkg/provider/chef/chef.go:118]:
```suggestion
	if len(secretKey) == 0 {
```

len already does nil check

File: pkg/provider/chef/chef.go [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+	}
+	databagName := ref.Key
+
+	if strings.Contains(databagName, "/") {
+		return nil, fmt.Errorf(errInvalidDataform)
+	}
+	getAllSecrets := make(map[string][]byte)
+	log.Info("fetching all items from databag:", databagName)
+	dataItems, err := providerchef.databagService.ListItems(databagName)
+	metrics.ObserveAPICall(ProviderChef, CallChefListDataBagItems, err)
```
Comment (Skarlso) [inline pkg/provider/chef/chef.go:233]: I think we would want this metric even if it failed in the first `IsNil` call, right?

File: pkg/provider/chef/chef.go [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+	_, err := providerchef.userService.Get(providerchef.clientName)
+	metrics.ObserveAPICall(ProviderChef, CallChefGetUser, err)
+	if err != nil {
+		return v1beta1.ValidationResultError, fmt.Errorf(errStoreValidateFailed)
+	}
+	return v1beta1.ValidationResultReady, nil
+}
+
+// GetAllSecrets Retrieves a map[string][]byte with the Databag names as key and the Databag's Items as secrets.
+// Retrives all DatabagItems of a Databag.
```
Comment (Skarlso) [inline pkg/provider/chef/chef.go:154]:
```suggestion
```

File: pkg/provider/chef/chef.go [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+	if databagName != "" && databagItem != "" {
+		return getSingleDatabagItem(providerchef, databagName, databagItem, ref.Property)
+	}
+
+	return nil, fmt.Errorf(errInvalidFormat)
+}
+
+func getSingleDatabagItem(providerchef *Providerchef, dataBagName, databagItemName, propertyName string) ([]byte, error) {
+	ditem, err := providerchef.databagService.GetItem(dataBagName, databagItemName)
+	metrics.ObserveAPICall(ProviderChef, CallChefGetDataBagItem, err)
```
Comment (Skarlso) [inline pkg/provider/chef/chef.go:183]: put this on the first line

File: pkg/provider/chef/chef.go (hunk 5 above)
Comment (Skarlso) [reply pkg/provider/chef/chef.go:233]: I understand that it logs the error, but this metric will no longer be available if chef is uninitialized.

Meh, maybe, this is fine. :D

**Review by Skarlso — CHANGES_REQUESTED** [review]
Generally looks okay. The branch needs a `go mod tidy` and a signed commit. :)

**Comment by sourav977** [comment]
The missed unsigned commits have been signed by me and my teammates but DCO check now complaining about other commits . @Skarlso Could you help me in resolving DCO check issue ?

**Comment by sourav977** [comment]
@Skarlso @sebagomez 
> Alternatively, take the diff, and create another PR and start off fresh if squashing is too complicated.

As suggested, we are closing this PR as we are unable to resolve unsigned commits , latest PR: https://github.com/external-secrets/external-secrets/pull/3127

**Pull request closed without merging**